ui/tests/unit/routes/vault/cluster/oidc-callback-test.js

```diff
@@ -207,4 +208,121 @@ module('Unit | Route | vault/cluster/oidc-callback', function (hooks) {
       'namespace is parsed correctly'
     );
   });
+
+  module('getParamsForCallback helper fn', function () {
+    test('it parses params correctly with regular inputs and no namespace', function (assert) {
+      const qp = {
+        state: 'my-state',
+        code: 'my-code',
+        path: 'oidc-path',
+      };
+      const searchString = `?code=my-code&state=my-state`;
+      const results = getParamsForCallback(qp, searchString);
+      assert.deepEqual(results, { source: 'oidc-callback', ...qp });
+    });
+
+    test('it parses params correctly regular inputs and namespace param', function (assert) {
+      const params = {
+        state: 'my-state',
+        code: 'my-code',
+        path: 'oidc-path',
+        namespace: 'my-namespace',
+      };
+      const results = getParamsForCallback(params, '?code=my-code&state=my-state&namespace=my-namespace');
+      assert.deepEqual(results, { source: 'oidc-callback', ...params });
+    });
+
+    test('it parses params correctly with regular inputs and namespace in state (unencoded)', function (assert) {
+      const searchString = '?code=my-code&state=my-state,ns=foo/bar';
+      const params = {
+        state: 'my-state,ns', // Ember parses the QP incorrectly if unencoded
+        code: 'my-code',
+        path: 'oidc-path',
+      };
+      const results = getParamsForCallback(params, searchString);
+      assert.deepEqual(results, {
+        source: 'oidc-callback',
+        ...params,
+        state: 'my-state',
+        namespace: 'foo/bar',
+      });
+    });
+
+    test('it parses params correctly with regular inputs and namespace in state (encoded)', function (assert) {
+      const qp = {
+        state: 'my-state,ns=foo/bar', // Ember parses the QP correctly when encoded
+        code: 'my-code',
+        path: 'oidc-path',
+      };
+      const searchString = `?code=my-code&state=${encodeURIComponent(qp.state)}`;
+      const results = getParamsForCallback(qp, searchString);
+      assert.deepEqual(results, { source: 'oidc-callback', ...qp, state: 'my-state', namespace: 'foo/bar' });
+    });
+
+    test('namespace in state takes precedence over namespace in route (encoded)', function (assert) {
+      const qp = {
+        state: 'my-state,ns=foo/bar',
+        code: 'my-code',
+        path: 'oidc-path',
+        namespace: 'other/ns',
+      };
+      const searchString = `?code=my-code&state=${encodeURIComponent(
+        qp.state
+      )}&namespace=${encodeURIComponent(qp.namespace)}`;
+      const results = getParamsForCallback(qp, searchString);
+      assert.deepEqual(results, {
+        source: 'oidc-callback',
+        ...qp,
+        state: 'my-state',
+        namespace: 'foo/bar',
+      });
+    });
+
+    test('namespace in state takes precedence over namespace in route (unencoded)', function (assert) {
```

Comment:
I haven't seen the namespace sent in the route like this! When does this happen?

Reply:
When the URL includes the namespace parameter that's the "in route" version. It might not be an actual case in the wild, but behaviorally the method overwrites the route namespace param with the one in state so I wanted to capture that behavior as expected

Reply:
Gotcha! Thanks for explaining 😄